### ISSUE TITLE
PKIX: CA Constraint after TrustAnchor

### ISSUE THREAD

**Issue by sbernard31** [eclipse/leshan#936]
(To better understand this point, you need to first read #935)

We are talking about Certificate Usage 0 "CA constraint" defined in  [RFC6698](https://tools.ietf.org/html/rfc6698#section-2.1.1) (used since LWM2M 1.1)

>   0 -- Certificate usage 0 is used to specify a CA certificate, or
      the public key of such a certificate, that MUST be found in any of
      the PKIX certification paths for the end entity certificate given
      by the server in TLS.  This certificate usage is sometimes
      referred to as "CA constraint" because it limits which CA can be
      used to issue certificates for a given service on a host.  The
      presented certificate MUST pass PKIX certification path
      validation, and a CA certificate that matches the TLSA record MUST
      be included as part of a valid certification path.  Because this
      certificate usage allows both trust anchors and CA certificates,
      the certificate might or might not have the basicConstraints
      extension present.

Generally in this use case the chains looks like this : 
```
Server -> CA (n) --> CA (n-1) --> ... --> CA 1  --> Root CA
```
(My understanding : this is possible to eventually have a chain without CA.)

A typical use case : 
```
Server -> CA 2 --> CA 1 --> Root CA
           ^        ^
           |        |
 CA Constraint     TrustAnchor
```
As  "CA constraint" can also be consider as a TrustAnchor and following #935 we will probably stop the PKIX validation at CA2. (From Server to CA2).

**The question is : What happens if CA Constraint is after another TrustAnchor .** 
E.g : 
```
Server -> CA 2 --> CA 1 --> Root CA
           ^        ^
           |        |
   TrustAnchor    CA Constraint
```

Following same logic we will  stop PKIX Validation to CA2 and we should also check that CA constraint (CA1 in this case) is in the certificate path. In this case CA1 is in the path but not really in the PKIX validation path which sounds not so good ?

**Comment by sbernard31** [comment]
Note that we the current implementation, It seems to me that  that "CA constraint" works pretty much the same as "Trust Anchor Assertion".

This comes from [this code](https://github.com/eclipse/leshan/pull/923/commits/6e96cee3e69ae7c6922946b56c88e1c2a868ff30#diff-22330e28a7d3c02e6f39f6db974b1345eb580153ac0943100226d9dd8164b049R156) where we choose to add the  "CA constraint" as a "trusted certificate"(TrustAnchor)

I suppose this is because of this RFC sentence (which is not so clear to me): 
> Because this certificate usage allows both trust anchors and CA certificates,
the certificate might or might not have the basicConstraints
extension present.

I'm not sure we should keep this as this bring issues describes above  AND it does not really solve any problem as if you want to use your certificate as trust anchors why not directly use the "Trust Anchor Assertion" usage ?

**Comment by sbernard31** [comment]
This should be implemented in #1157.

**Comment by sbernard31** [comment]
(integrated in master)
